subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c

```diff
@@ -146,140 +146,132 @@ gst_msdkdec_get_oldest_frame (GstVideoDecoder * decoder)
   }
 
-  if (!frame->output_buffer) {
-    /* Free un-unsed msdk surfaces firstly, hence the associated mfx
-     * surfaces will be moved from used list to available list */
-    gst_msdkdec_free_unlocked_msdk_surfaces (thiz);
+  mem = gst_buffer_peek_memory (out_buffer, 0);
+  msdk_surface = g_slice_new0 (GstMsdkSurface);
+  msdk_surface->from_qdata = FALSE;
 
-    flow = gst_video_decoder_allocate_output_frame (decoder, frame);
-    if (flow != GST_FLOW_OK) {
-      gst_video_codec_frame_unref (frame);
-      return flow;
-    }
+  if ((mfx_surface = gst_mini_object_get_qdata (GST_MINI_OBJECT_CAST (mem),
+              GST_MSDK_FRAME_SURFACE))) {
+    msdk_surface->surface = mfx_surface;
+    msdk_surface->from_qdata = TRUE;
+  } else {
+    GST_ERROR ("No available surfaces");
+    g_slice_free (GstMsdkSurface, msdk_surface);
+    return NULL;
+  }
+#else
+  if (!gst_buffer_pool_is_active (thiz->pool) &&
+      !gst_buffer_pool_set_active (thiz->pool, TRUE)) {
+    GST_ERROR_OBJECT (thiz, "Failed to activate buffer pool");
+    return NULL;
   }
 
-  *buffer = gst_buffer_ref (frame->output_buffer);
-  gst_buffer_replace (&frame->output_buffer, NULL);
-  gst_video_codec_frame_unref (frame);
+  if ((gst_buffer_pool_acquire_buffer (thiz->pool, &out_buffer, 0))
+      != GST_FLOW_OK) {
+    GST_ERROR_OBJECT (thiz, "Failed to allocate output buffer");
+    return NULL;
+  }
 
-  return GST_FLOW_OK;
-}
+  GstVideoCodecState *output_state =
+      gst_video_decoder_get_output_state (GST_VIDEO_DECODER (thiz));
 
-static MsdkSurface *
-get_surface (GstMsdkDec * thiz, GstBuffer * buffer)
-{
-  MsdkSurface *i;
-  GstVideoCodecState *output_state = NULL;
-  gboolean success;
+  msdk_surface =
+      gst_msdk_import_sys_mem_to_msdk_surface (out_buffer, output_state->info);
 
-  i = g_slice_new0 (MsdkSurface);
+  gst_video_codec_state_unref (output_state);
+#endif
 
-  if (gst_msdk_is_msdk_buffer (buffer)) {
-    i->surface = gst_msdk_get_surface_from_buffer (buffer);
-    i->buf = buffer;
+  if (msdk_surface) {
+    msdk_surface->buf = out_buffer;
   } else {
-    /* Confirm to activate the side pool */
-    if (!gst_buffer_pool_is_active (thiz->pool) &&
-        !gst_buffer_pool_set_active (thiz->pool, TRUE)) {
-      g_slice_free (MsdkSurface, i);
-      return NULL;
-    }
-
-    if (!gst_video_frame_map (&i->copy, &thiz->non_msdk_pool_info, buffer,
-            GST_MAP_WRITE))
-      goto failed_unref_buffer;
-
-    if (gst_buffer_pool_acquire_buffer (thiz->pool, &buffer,
-            NULL) != GST_FLOW_OK)
-      goto failed_unmap_copy;
-
-    i->surface = gst_msdk_get_surface_from_buffer (buffer);
-    i->buf = buffer;
-
-    output_state =
-        gst_video_decoder_get_output_state (GST_VIDEO_DECODER (thiz));
-    success =
-        gst_video_frame_map (&i->data, &output_state->info, buffer,
-        GST_MAP_READWRITE);
-    gst_video_codec_state_unref (output_state);
-    if (!success)
-      goto failed_unref_buffer2;
+    return NULL;
```

Comment:
I think maybe you should put this in the system memory path.

Reply:
As we discussed before, we don't directly use system memory for decoder, so the decoder allocated outbuf is always video memory. We only consider doing gpu2cpu copy when pushing the outbuf forward to downstream, this process can be seen in gst_mskdec_finish_task().